## Week 5 Homework 

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHV 2019-10 data found here. [FHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz)


### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?

> [!NOTE]
> To install PySpark follow this [guide](https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/05-batch/setup/pyspark.md)

In [1]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

24/03/04 01:11:55 WARN Utils: Your hostname, SPCR.local resolves to a loopback address: 127.0.0.1; using 192.168.0.108 instead (on interface en0)
24/03/04 01:11:55 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/03/04 01:11:55 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


# Q1 - Answer 

In [60]:
#pyspark.__version__
spark.version

'3.5.1'

### Question 2: 

**FHV October 2019**

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

- 1MB
- 6MB
- 25MB
- 87MB

In [39]:
year = 2019
month = 10
type = 'fhv'

In [40]:
!sh download_data.sh {type} {year} {month}

downloading https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz to data/raw/fhv/2019/10/fhv_tripdata_2019_10.csv.gz
--2024-03-04 01:48:59--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240304T004900Z&X-Amz-Expires=300&X-Amz-Signature=1552a6a9cf5f10eaa90c9abda252c4a7abac22b905e30fe18bf9229c6ee60505&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=appli

In [41]:
import pandas as pd

df_fhv_pd = pd.read_csv(f'data/raw/{type}/{year}/{month:02d}/{type}_tripdata_{year}_{month:02d}.csv.gz', nrows=100)

list = spark.createDataFrame(df_fhv_pd).schema
f = open('_fhv_schema_def_', 'w')
f.write(str(list))

371

In [43]:
from pyspark.sql import types

fhv_schema = types.StructType([
     types.StructField('dispatching_base_num', types.StringType(), True)
    ,types.StructField('pickup_datetime', types.TimestampType(), True)
    ,types.StructField('dropoff_datetime', types.TimestampType(), True)
    ,types.StructField('PULocationID', types.IntegerType(), True)
    ,types.StructField('DOLocationID', types.IntegerType(), True)
    ,types.StructField('SR_Flag', types.DoubleType(), True)
    ,types.StructField('Affiliated_base_number', types.StringType(), True)
])

in_fhv_path = f'data/raw/{type}/{year}/{month:02d}/'


df_fhv = spark.read \
    .option("header", "true") \
    .schema(fhv_schema) \
    .csv(in_fhv_path)

df_fhv.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropoff_datetime: timestamp (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- SR_Flag: double (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [44]:
out_fhv_path = f'data/parquet/{type}/{year}/{month:02d}/'

df_fhv \
    .repartition(6) \
    .write \
    .parquet(out_fhv_path)

# Q2 - Answer

In [59]:
!ls -lh {out_fhv_path} | grep '.parquet' 

-rw-r--r--  1 peter  staff   6.4M Mar  4 01:51 part-00000-15cb5243-f655-4d15-bce3-d4f0c0b3dbb2-c000.snappy.parquet
-rw-r--r--  1 peter  staff   6.3M Mar  4 01:51 part-00001-15cb5243-f655-4d15-bce3-d4f0c0b3dbb2-c000.snappy.parquet
-rw-r--r--  1 peter  staff   6.4M Mar  4 01:51 part-00002-15cb5243-f655-4d15-bce3-d4f0c0b3dbb2-c000.snappy.parquet
-rw-r--r--  1 peter  staff   6.4M Mar  4 01:51 part-00003-15cb5243-f655-4d15-bce3-d4f0c0b3dbb2-c000.snappy.parquet
-rw-r--r--  1 peter  staff   6.4M Mar  4 01:51 part-00004-15cb5243-f655-4d15-bce3-d4f0c0b3dbb2-c000.snappy.parquet
-rw-r--r--  1 peter  staff   6.4M Mar  4 01:51 part-00005-15cb5243-f655-4d15-bce3-d4f0c0b3dbb2-c000.snappy.parquet


### Question 3: 

**Count records** 

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

- 108,164
- 12,856
- 452,470
- 62,610

> [!IMPORTANT]
> Be aware of columns order when defining schema

# Q3 - Answer

### Question 4: 

**Longest trip for each day** 

What is the length of the longest trip in the dataset in hours?

- 631,152.50 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

# Q4 - Answer

### Question 5: 

**User Interface**

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- 4040
- 8080

# Q5 - Answer

### Question 6: 

**Least frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?</br>

- East Chelsea
- Jamaica Bay
- Union Sq
- Crown Heights North

# Q6 - Answer